sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java

```diff
@@ -1700,43 +1699,31 @@ public DefaultRunner(Logger standardLogger, EngineLogger engineLogger) {
             this.engineLogger = Objects.requireNonNull(engineLogger);
         }
 
-        public List<Exception> getSwallowedExceptions() {
-            return ImmutableList.copyOf(this.swallowedExceptions);
-        }
-
-        @Override
-        public <T extends Stack> CompletableFuture<Integer> runAsync(Supplier<T> stackFactory) {
-            try {
-                var stack = stackFactory.get();
-                var stackInternal = Internal.from(stack);
-                // Stack doesn't call RegisterOutputs, so we register them on its behalf.
-                stackInternal.registerPropertyOutputs();
-                registerTask(String.format("runAsync: %s, %s", stack.getResourceType(), stack.getResourceName()),
-                        Internal.of(stackInternal.getOutputs()).getDataAsync());
-            } catch (Exception ex) {
-                return handleExceptionAsync(ex);
-            }
-
-            return whileRunningAsync();
-        }
-
-        @Override
-        public CompletableFuture<Integer> runAsyncFuture(Supplier<CompletableFuture<Map<String, Output<?>>>> callback) {
-            return runAsyncFuture(callback, StackOptions.Empty);
-        }
-
         @Override
-        public CompletableFuture<Integer> runAsyncFuture(Supplier<CompletableFuture<Map<String, Output<?>>>> callback, StackOptions options) {
-            try {
-                var stack = StackInternal.of(callback, options);
-                // no outputs to register here
-                registerTask(String.format("runAsyncFuture: %s, %s", stack.getResourceType(), stack.getResourceName()),
-                        Internal.of(Internal.from(stack).getOutputs()).getDataAsync());
-            } catch (Exception ex) {
-                return handleExceptionAsync(ex);
-            }
-
-            return whileRunningAsync();
+        public <T> CompletableFuture<Result<T>> runAsync(Supplier<T> callback) {
+            var valueFuture = CompletableFuture.supplyAsync(callback);
+            // run the callback asynchronously in the context of the error handler
+            registerTask("DefaultRunner#runAsync", valueFuture);
+            // loop starts after the callback
+            return valueFuture
+                    .thenCompose(__ -> whileRunningAsync())
+                    .handle((code, throwable) -> {
+                        if (throwable != null) {
+                            return handleExceptionAsync(throwable).thenApply(errorCode ->
+                                    new Result<>(
+                                            errorCode,
+                                            ImmutableList.copyOf(this.swallowedExceptions),
+                                            Optional.ofNullable(valueFuture.join()) // the future is already complete
```

Comment:
It completed with an exception though, why do you think it would have a value? Naively I'd expect `.join()` to throw here but I'm not sure, better try it out and see.

Reply:
Hmm, that's a good point, does it throw, it should throw on `join`.
I can't exclude a situation where both the value and an exception is present, but not sure if it is a useful state. 
Maybe simplifying this to always return Optional.empty() could be a better option, but I was worried this will make some tests impossible/difficult to write.

Reply:
After moving error handling to one place, the join is no longer needed. I'm glad this small hack of mine got deleted :)